service/controller/resource/masters/create_master_instances_upgrading.go

```diff
@@ -57,143 +58,95 @@ func (r *Resource) masterInstancesUpgradingTransition(ctx context.Context, obj i
 		} else {
 			r.Logger.LogCtx(ctx, "level", "debug", "message", "processing master VMSSs")
 
-			ws, err := r.nextInstance(ctx, cr, allMasterInstances, drainerConfigs, key.MasterInstanceName, versionValue)
-			if err != nil {
-				return "", microerror.Mask(err)
+			// Ensure that all VM instances are in Successful state before proceeding with reimaging.
+			for _, vm := range allMasterInstances {
+				if vm.ProvisioningState != nil && !key.IsSucceededProvisioningState(*vm.ProvisioningState) {
+					r.Logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("master instance %#q is not in successful provisioning state: %#q", key.MasterInstanceName(cr, *vm.InstanceID), *vm.ProvisioningState))
+					r.Logger.LogCtx(ctx, "level", "debug", "message", "cancelling resource")
+					return currentState, nil
+				}
 			}
 
-			err = r.updateInstance(ctx, cr, ws.InstanceToUpdate(), key.MasterVMSSName, key.MasterInstanceName)
-			if err != nil {
-				return "", microerror.Mask(err)
-			}
-			if ws.InstanceToDrain() != nil {
-				err = r.CreateDrainerConfig(ctx, key.ClusterID(&cr), key.ClusterAPIEndpoint(cr), key.MasterInstanceName(cr, *ws.InstanceToDrain().InstanceID))
+			desiredVersion := project.Version()
+			for _, vm := range allMasterInstances {
+				instanceName := key.MasterInstanceName(cr, *vm.InstanceID)
+				instanceVersion, ok := versionValue[instanceName]
+				if !ok {
+					continue
+				}
+				if desiredVersion == instanceVersion {
+					continue
+				}
+
+				masterUpgradeInProgress = true
+
+				// Ensure that VM has latest VMSS configuration (includes ignition template etc.).
+				if !*vm.VirtualMachineScaleSetVMProperties.LatestModelApplied {
```

Comment:
We need to test this very carefully in germany west central.
We used to rely on `LatestModelApplied` before and we had to get rid of it because it was not working as expected in GWC.

Reply:
We did rely on that before as well:
https://github.com/giantswarm/azure-operator/pull/1142/files#diff-aa9eaeec1e181519c60c3144c3c2aca5687f17dbe148b50506cc92ad18b8b2deL345-L351

Reply:
But in case it doesn't make the node reboot or so, there's the catch w/ reimaging.